### ISSUE TITLE
Exception raised with numexpr v2.8.6

### ISSUE THREAD

**Issue by t20100** [silx-kit/pyFAI#1947]
With `numexpr` v2.8.6`, `from pyFAI.azimuthalIntegrator import AzimuthalIntegrator` raises an exception:

`ValueError: Expression 4.0e-9*π/λ*sin(arctan2(sqrt(x * x + y * y), z)/2.0) has forbidden control characters.`

<details> <summary>Logs:</summary>

```
File ~/venvs/py310/lib/python3.10/site-packages/pyFAI/units.py:230, in <module>
    211 register_radial_unit("2th_deg", scale=180.0 / numpy.pi,
    212                      center="twoThetaArray",
    213                      delta="delta2Theta",
   (...)
    217                      short_name=r"2\theta",
    218                      unit_symbol="deg")
    220 register_radial_unit("2th_rad",
    221                      center="twoThetaArray",
    222                      delta="delta2Theta",
   (...)
    227                      short_name=r"2\theta",
    228                      unit_symbol="rad")
--> 230 register_radial_unit("q_nm^-1",
    231                      center="qArray",
    232                      delta="deltaQ",
    233                      scale=1.0,
    234                      label=r"Scattering vector $q$ ($nm^{-1}$)",
    235                      equation=eq_q,
    236                      formula=formula_q,
    237                      short_name="q",
    238                      unit_symbol="nm^{-1}")
    240 register_radial_unit("q_A^-1",
    241                      center="qArray",
    242                      delta="deltaQ",
   (...)
    247                      short_name="q",
    248                      unit_symbol=r"\AA^{-1}")
    250 register_radial_unit("d*2_A^-2",
    251                      center="rd2Array",
    252                      delta="deltaRd2",
   (...)
    257                      short_name="d^{*2}",
    258                      unit_symbol=r"\AA^{-2}")

File ~/venvs/py310/lib/python3.10/site-packages/pyFAI/units.py:149, in register_radial_unit(name, scale, label, equation, formula, center, corner, delta, short_name, unit_symbol)
    147 def register_radial_unit(name, scale=1, label=None, equation=None, formula=None,
    148                          center=None, corner=None, delta=None, short_name=None, unit_symbol=None):
--> 149     RADIAL_UNITS[name] = Unit(name, scale, label, equation, formula, center,
    150                               corner, delta, short_name, unit_symbol)

File ~/venvs/py310/lib/python3.10/site-packages/pyFAI/units.py:110, in Unit.__init__(self, name, scale, label, equation, formula, center, corner, delta, short_name, unit_symbol)
    108 if "π" in formula:
    109     signature.append(("π", numpy.float64))
--> 110 ne_formula = numexpr.NumExpr(formula, signature)
    112 def ne_equation(x, y, z=None, wavelength=None, ne_formula=ne_formula):
    113     π = numpy.pi

File ~/venvs/py310/lib/python3.10/site-packages/numexpr/necompiler.py:639, in NumExpr(ex, signature, sanitize, **kwargs)
    637 _frame_depth = 1
    638 context = getContext(kwargs, _frame_depth=_frame_depth)
--> 639 threeAddrProgram, inputsig, tempsig, constants, input_names = precompile(ex, signature, context, sanitize=sanitize)
    640 program = compileThreeAddrForm(threeAddrProgram)
    641 return interpreter.NumExpr(inputsig.encode('ascii'),
    642                            tempsig.encode('ascii'),
    643                            program, constants, input_names)

File ~/venvs/py310/lib/python3.10/site-packages/numexpr/necompiler.py:576, in precompile(ex, signature, context, sanitize)
    573 input_order = [name for (name, type_) in signature]
    575 if isinstance(ex, str):
--> 576     ex = stringToExpression(ex, types, context, sanitize)
    578 # the AST is like the expression, but the node objects don't have
    579 # any odd interpretations
    581 ast = expressionToAST(ex)

File ~/venvs/py310/lib/python3.10/site-packages/numexpr/necompiler.py:281, in stringToExpression(s, types, context, sanitize)
    279     no_whitespace = re.sub(r'\s+', '', s)
    280     if _blacklist_re.search(no_whitespace) is not None:
--> 281         raise ValueError(f'Expression {s} has forbidden control characters.')
    283 old_ctx = expressions._context.get_current_context()
    284 try:

ValueError: Expression 4.0e-9*π/λ*sin(arctan2(sqrt(x * x + y * y), z)/2.0) has forbidden control characters.
```

</details>

This looks to come from `4.0e-9`:

- This raises the exception `numexpr.NumExpr("4.0e-9*π/λ*sin(arctan2(sqrt(x * x + y * y), z)/2.0)")`.
- This does not: `numexpr.NumExpr("4.e-9*π/λ*sin(arctan2(sqrt(x * x + y * y), z)/2.0)")`

**Comment by t20100** [comment]
Related to https://github.com/pydata/numexpr/issues/449

**Comment by loichuder** [comment]
Encountered this as well.

Note that it is possible to disable sanitization but it doesn't fix the issue (https://github.com/pydata/numexpr/blob/master/RELEASE_NOTES.rst#changes-from-285-to-286)...

**Comment by newville** [comment]
I'm seeing this as well.  A parsing error like that from numexpr is pretty concerning.  Like, it is failing to parse valid numerical literals in scientific notation?  Yikes.  

Why is pyFAI.units even using numexpr at all?  There are only a few expressions needed, all of them are fully encapsulated in a single line of Python.  

Please convert to plain Python or Cython.

**Comment by kif** [comment]
Believe me, error messages have arrived to me the whole day on this topic.
`numexpr` has been there for several years and pyFAI runs perfectly without it ... but not when it is installed and non  functional.

> Why is pyFAI.units even using numexpr at all? There are only a few expressions needed, all of them are fully encapsulated in a single line of Python.

The main reason is to allow users to implement their own radial/azimuthal units if they feel like. You can easily implement log(2th) if you wish to integrate in this radial unit. There is a tutorial on the topic.

http://www.silx.org/doc/pyFAI/dev/usage/tutorial/LogScale/Guinier.html

> Please convert to plain Python or Cython.

The numexpr code is only used when the cython one is not available. Both are validated against the pure numpy implementation and checked in the test. numexpr is much faster than numpy to initialize the geometry.

**Comment by newville** [comment]
@kif Thanks, but I do have Cython installed and still get 

```
Python 3.10.10 | packaged by conda-forge | (main, Mar 24 2023, 20:17:34) [Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import cython
>>> cython.__version__
'3.0.2'
>>> import numexpr
>>> numexpr.__version__
'2.8.6'
>>> import pyFAI
>>> pyFAI._version.version
'2023.9.0'
>>> import pyFAI.units
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../lib/python3.10/site-packages/pyFAI/units.py", line 230, in <module>
    register_radial_unit("q_nm^-1",
  File ".../lib/python3.10/site-packages/pyFAI/units.py", line 149, in register_radial_unit
    RADIAL_UNITS[name] = Unit(name, scale, label, equation, formula, center,
  File ".../lib/python3.10/site-packages/pyFAI/units.py", line 110, in __init__
    ne_formula = numexpr.NumExpr(formula, signature)
  File ".../lib/python3.10/site-packages/numexpr/necompiler.py", line 639, in NumExpr
    threeAddrProgram, inputsig, tempsig, constants, input_names = precompile(ex, signature, context, sanitize=sanitize)
  File ".../lib/python3.10/site-packages/numexpr/necompiler.py", line 576, in precompile
    ex = stringToExpression(ex, types, context, sanitize)
  File ".../lib/python3.10/site-packages/numexpr/necompiler.py", line 281, in stringToExpression
    raise ValueError(f'Expression {s} has forbidden control characters.')
ValueError: Expression 4.0e-9*π/λ*sin(arctan2(sqrt(x * x + y * y), z)/2.0) has forbidden control characters.
>>>
```

I do understand that numexpr can give high performance and is very flexible, but I could also imagine that you just allow users to write a Python function for specialized azimuthal units.  Maybe that is too slow?

But also: a mature library explicitly designed to parse and evaluate numerical Python expressions failing to parse literal numbers in scientific notation is kind of amazing.

**Comment by kif** [comment]
@newville, I picked `numexpr` at that time because it offers the ability to parse mathematical expression, preventing injection of malformed/harmful python code.  If you have a look at the goniometer calibration, `numexpr` is at the core of it. Its usage in `pyFAI.units` is just the cherry on top of the cake.

On the other hand, I totally agree with you that this is not an acceptable bug from numexpr. @t20100 suggested a solution  (i.e. fix the string parser) and we are all waiting for the bug-fix release.

**Comment by kif** [comment]
I did try to make a `numexpr` version which patches the call to append `validate=False` but it fails: numexpr is doing too many clever things with locals() and globals().

**Comment by newville** [comment]
@kif @t20100 Thanks, and yes, I completely understand the choice.  I there will be a quick fix for this and maybe that is sufficient.  But, it might also be a good time to reassess having that dependency ;).  Thanks again, and Cheers!

**Comment by kif** [comment]
closed by #1950
